smolyay/basis.py

```diff
@@ -0,0 +1,232 @@
+import abc
+import math
+
+import numpy
+
+class BasisFunction(abc.ABC):
+    """Basis function for interpolating data.
+
+     A one-dimensional basis function is defined on the domain
+     :math:`[-1,1]`. The function defines the :attr:`points` at
+     which it should be sampled within this interval for interpolation.
+     The function also has an associated :meth:`__call__` method
+     for evaluating it at a point within its domain.
+
+
+    """
+
+    def __init__(self):
+        self._points = []
+
+    @property
+    @abc.abstractmethod
+    def points(self):
+        """list: Sampling points for interpolation."""
+        pass
+
+
+    @abc.abstractmethod
+    def __call__(self,x):
+        """Evaluate the basis function.
+
+        Parameters
+        ----------
+        x : float
+            One-dimensional point.
+
+        Returns
+        -------
+        float
+            Value of basis function.
+        """
+        pass
+
+
+class ChebyshevFirstKind(BasisFunction):
+    r"""Chebyshev polynomial of the first kind.
+
+    The Chebyshev polynomial of degree *n* is defined by the
+    recursive relationship:
+
+    ..math::
+
+        T_0(x) = 1
+        T_1(x) = x
+        T_{n+1}(x) = 2x T_n(x) - T_{n-1}(x)
+
+    The :attr:`points` for this polynomial are the extrema on the domain
+    :math:`[-1,1]`:
+   
+    ..math::
+
+        x_i^* = -\cos(\pi i/n), i = 0,...,n
+
+    Parameters
+    ----------
+    n : int
+        Degree of the Chebyshev polynomial.
+    """
+
+    def __init__(self,n):
+        super().__init__()
+        self._n = n
+        self._points = list(
+                numpy.polynomial.chebyshev.chebpts2(n+1)) if n > 0 else [0]
+
+    @property
+    def points(self):
+        """list: Sampling points at extrema of polynomial."""
+        if len(self._points) == 0:
+            self._compute_points()
+        return self._points
+
+    @property
+    def n(self):
+        """degree of polynomial"""
+        return self._n
+
+    def __call__(self,x):
```

Comment:
No, because the inputs for the call method are not expected to always be on the range [-1,1]. For example, If the input space for the real functions parameters aren't on that range (like on the branin function the surrogate was matching it on the range of [-5,10] for x1 and [-5,15] for x2).

Reply:
See comment above. It is the responsibility of other classes to shift and scale their inputs to the relevant domain. Your and Mohammadreza’s code did this before.

Reply:
I am looking at my code that I made at the beginning of the summer and the Chebyshev polynomial is definitely called outside of that domain. Mohammadreza's does the same thing.

Reply:
Then that code was incorrect. The real function may have a different domain, but the points/inputs need to be scaled from/to [-1,1] before the Chebyshev polynomial is called. I believe Dr. K explained this to us?

Reply:
Yeah, that appears to be correct with the literature.

Reply:
It seems like this is resolved, but yes we typically have two versions of the grid points. One on the domain of [-1 1] in all dimensions, which is used to sample the Smolyak polynomial. The other version of the grid points are those where the [-1 1] domain is transformed to that of the problem at hand, which is used to sample the model that is being approximated.